feat!: Add automatic retrying of 429/502/503

judofyr opened the pull request:
This is a bit of a bigger, and technically breaking, change which introduces automatic retrying of 429 errors (i.e. rate limited). I'm opening it here to start a discussion about how to progress this forward. I believe this would be very useful for the majority of the users of the client and should be enabled by default.

judofyr committed:
feat!: add automatic retrying of 429, 502, 503

BREAKING CHANGE: Client will now automatically retry all GET/HEAD
requests as well as queries if the server responds with a 429, 502 or
503 status code - as well as on socket/dns errors. Previously, the
client would immediately throw an error. If you have application-level
retry code, you should either disable the retrying in the client
by passing `{maxRetries: 0}`, or remove the custom retry code and
potentially alter the `retryDelay` and `maxRetries` options to match
your wanted behavior.

stipsan reviewed (commented):
could you change the PR title to: feat!: add automatic retry on 429

Just so that semantic release bumps the right version and generates appropriate release notes. We can do this major release together with the useCdn one so we don't have two breaking releases in short succession

judofyr commented:
> could you change the PR title to: feat!: add automatic retry on 429

It's enough with the PR title? Or should the commit also have the `!`?

stipsan commented:
Commit title as well

On Tue, 18 Apr 2023 at 14:13, Magnus Holm ***@***.***> wrote:

> could you change the PR title to: feat!: add automatic retry on 429
>
> It's enough with the PR title? Or should the commit also have the !?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sanity-io/client/pull/199#issuecomment-1512988915>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUAPL6WC4DZTMESFWKY6TXB2AN5ANCNFSM6AAAAAAXCP5QFQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

judofyr commented:
- Changed the commit message to `feat!`
- Added retrying of 502 and 503 as well.
- Added more comment explaining why we retry.

rexxars commented on src/http/request.ts:
```diff
@@ -27,8 +27,39 @@ const printWarnings = {
+          err.response &&
+          (err.response.statusCode === 429 ||
+            err.response.statusCode === 502 ||
+            err.response.statusCode === 503)
```
There is a case to be made for also retrying other non-response errors. For instance, temporary socket errors, DNS resolve failures etc. Normally these wouldn't be retried if the request is a POST-request - but we could retry POST _queries_ at least.

Additionally, DNS lookup failures would prevent a request from even going out, so we could also retry those if the URL is still `*api.sanity.io`/`*apicdn.sanity.io`, since those _should_ always resolve. I've seen multiple reports of DNS lookup errors being reported, and I assume it is a temporary resolve failure on the client side causing it since the alternative would show pretty rapidly on our charts (and reported issues).

Here are some common error codes: https://nodejs.org/api/errors.html#common-system-errors

judofyr replied:
We're already using `is-retry-allowed` in get-it which I believe should support this, but it's not applied to POST. Maybe the "proper" way to support this is to support `options.isSafe = true` (or maybe `options.isIdempotent = true` since we can actually also retry these?), and then have the default `shouldRetry` check for this instead?

```js
function shouldRetry(…) {
  const isSafe = options.isSafe || options.method === 'GET' || options.method === 'HEAD' // remain backwards compatible
  if (!isSafe) return false
}
```

rexxars commented on src/index.ts:
```diff
@@ -11,12 +11,19 @@ export * from './SanityClient'
 
 // Set the http client to use for requests, and its environment specific middleware
-const httpRequest = defineHttpRequest(envMiddleware)
+const httpRequest = defineHttpRequest(envMiddleware, {maxRetries: 0})
```
What is the rationale/intended behavior of `maxRetries: 0` here?
[This line](https://github.com/sanity-io/get-it/blob/main/src/middleware/retry/shared-retry.ts#L8) in get-it leads me to believe that it would result in the _actual_ max retries being set to 5, which may or may not be what you want, depending on the intention.

judofyr replied:
The _intention_ was that the requester would get no retries in order to make it "less" of a breaking change. I'm not sure if this makes sense though. It's maybe more consistent to default to 5 there as well?

rexxars commented on src/http/request.ts:
```diff
@@ -27,8 +27,39 @@ const printWarnings = {
+    retry({
+      retryDelay: retryDelay as () => number, // This option is typed incorrectly in get-it.
+      maxRetries,
+      shouldRetry(err, attempt, options) {
```
Could we export this for people that want to reuse it in their custom `shouldRetry` functions?

judofyr replied:
Currently `shouldRetry` _isn't_ exposed as a public options. The intention was that this is something which shouldn't be configured by an end-user as this is determined by what the Sanity backend supports (and hence should be "hard coded" in the client).

rexxars reviewed (changes requested):
- This looks great - will be really nice to get those temporary errors retried. I think this is unlikely to "solve" rate limiting (with such low first-retry timeouts), but also probably won't make it worse. I Have A Branch™ that implements actual rate-limiting that I'd like to finish up at some point, but we can leave that for a later time.

- Maybe we should provide a way to disable the retrying? Passing `{maxRetries: 0}` will actually end up setting it to `5` currently (unfortunate choice, I guess we might want to fix this in get-it).

- I would prefer if we also added an explicit `BREAKING CHANGE:` explanation to the commit(s), in order for explain it in a little more detail for automatic release notes. For instance:
```
BREAKING CHANGE: Client will now automatically retry all GET/HEAD requests as
well as queries if the server responds with a 429, 502 or 503 status code - as
well as on socket/dns errors. Previously, the client would immediately throw
an error. If you have application-level retry code, you should either disable
the retrying in the client by passing `{maxRetries: 0}`, or remove the custom
retry code and potentially alter the `shouldRetry` and `maxRetries` options to
match your wanted behavior.
```

judofyr commented:
Okay, I've made the following changes:

- Made it possible to disable the retrying logic by specifying `maxRetries: 0`.
- Updated commit message.
- Removed the `retry` that we used in the Node middleware. (Wasn't aware of this!)
- Changed the exported `requester` to have the same retry logic.

I'd prefer to merge this now and then rather implement "retry also on DNS errors for POST requests" in a later (patch) release as I think this is best solved by adding support for it in `get-it`.

judofyr committed:
fix: keep index.browser.ts in sync with index.ts

stipsan commented on src/index.ts:
```diff
@@ -11,12 +11,19 @@ export * from './SanityClient'
 
 // Set the http client to use for requests, and its environment specific middleware
-const httpRequest = defineHttpRequest(envMiddleware)
+const httpRequest = defineHttpRequest(envMiddleware, {})
 /** @public */
 export const requester = httpRequest.defaultRequester
```
Out of scope for this PR, but it does seem like we should export `defineHttpRequest` instead of `requester`. I'm not sure what use there is for `requester` anymore. What do you think @rexxars?

rexxars replied:
There are a few usages internally, but I'm sure they could be done differently: https://github.com/search?q=org%3Asanity-io+requester&type=code

stipsan reviewed (approved):
All good!
